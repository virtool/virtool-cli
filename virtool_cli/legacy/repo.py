"""Repo level checks for legacy repositories."""

from pathlib import Path

from virtool_cli.legacy.utils import iter_legacy_otus
from virtool_cli.utils.console import console


def check_unique_accessions(path: Path):
    """Validate that all accession fields are unique."""
    accessions = set()

    for otu in iter_legacy_otus(path / "src"):
        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                if accession := sequence.get("accession"):
                    if accession in accessions:
                        console.log(f"Found non-unique accession: {accession}")
                    else:
                        accessions.add(accession)


def check_unique_ids(path: Path):
    """Validate that resource IDs are unique.

    * All OTU _id fields are unique.
    * All isolate id fields are unique.
    * All sequence _id fields are unique.
    """
    otu_ids = []
    isolate_ids = []
    sequence_ids = []

    for otu in iter_legacy_otus(path / "src"):
        otu_ids.append(otu["_id"])

        for isolate in otu["isolates"]:
            isolate_ids.append(isolate["id"])

            for sequence in isolate["sequences"]:
                sequence_ids.append(sequence["_id"])

    if len(otu_ids) != len(set(otu_ids)):
        console.log("Found non-unique OTU ids")

    if len(isolate_ids) != len(set(isolate_ids)):
        console.log("Found non-unique isolate ids")

    if len(sequence_ids) != len(set(sequence_ids)):
        console.log("Found non-unique sequence ids")


def check_unique_otu_abbreviations_and_names(path: Path):
    """Validate that OTU abbreviations and names are unique."""
    abbreviations = set()
    names = set()

    for otu in iter_legacy_otus(path / "src"):
        abbreviation = otu["abbreviation"]

        if abbreviation:
            if abbreviation in abbreviations:
                console.log(f"Found non-unique OTU abbreviation: {abbreviation}")
            else:
                abbreviations.add(abbreviation)

        name = otu["name"].lower()

        if name in names:
            console.log(f"Found non-unique OTU name: {name}")
        else:
            names.add(name)
