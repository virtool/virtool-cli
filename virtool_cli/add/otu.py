import json
from pathlib import Path
import asyncio
import structlog
from urllib.error import HTTPError

from virtool_cli.utils.logging import DEFAULT_LOGGER, DEBUG_LOGGER
from virtool_cli.utils.reference import is_v1, generate_otu_dirname, get_unique_otu_ids
from virtool_cli.utils.id_generator import generate_unique_ids
from virtool_cli.utils.ncbi import fetch_taxonomy_record
from virtool_cli.catalog.listings import add_new_listing

OTU_KEYS = ["_id", "name", "abbreviation", "schema", "taxid"]

base_logger = structlog.get_logger()


def run(
    taxid: int,
    src_path: Path,
    catalog_path: Path,
    debugging: bool = False,
):
    """
    CLI entry point for virtool_cli.add.otu module

    Requests updates for a single OTU directory
    Searches the catalog for a matching catalog listing and requests new updates if it finds one.

    :param taxid:
    :param src_path: Path to a reference directory
    :param catalog_path: Path to a catalog directory
    :param debugging: Enables verbose logs for debugging purposes
    """
    structlog.configure(wrapper_class=DEBUG_LOGGER if debugging else DEFAULT_LOGGER)
    logger = base_logger.bind(src=str(src_path))

    if is_v1(src_path):
        logger.critical(
            'Reference folder "src" is a deprecated v1 reference.'
            + 'Run "virtool ref migrate" before trying again.'
        )
        return

    logger.info(f"Creating OTU directory for NCBI Taxonomy uid {taxid}")

    asyncio.run(add_otu(taxid=taxid, src_path=src_path, catalog_path=catalog_path))


async def add_otu(taxid: int, src_path: Path, catalog_path: Path):
    """
    Fetch NCBI Taxonomy data for a given UID and write the OTU to
    reference directory src and a listing to the catalog

    :param taxid:
    :param src_path:
    :param catalog_path:
    """
    logger = base_logger.bind(taxon_id=taxid)

    try:
        taxonomy_data = await fetch_taxonomy_record(taxon_id=taxid)
        if not taxonomy_data:
            logger.error("Could not find a record under this taxon ID on NCBI Taxonomy")
            return

    except HTTPError:
        logger.error(
            "Could not download this data. Please wait a few seconds before trying again"
        )
        return

    taxid_matches = list(catalog_path.glob(f"{taxid}--*.json"))
    if taxid_matches:
        match = taxid_matches.pop()
        logger.error(
            "An OTU with this taxon ID already exists in the reference database.",
            listing=match.name,
        )
        return

    unique_otu_ids = await get_unique_otu_ids(src_path)
    new_id = generate_unique_ids(1, excluded=unique_otu_ids).pop()

    new_otu = generate_otu(taxonomy_data, new_id)
    if not new_otu:
        logger.error("Invalid data from NCBI Taxonomy")

    logger.debug(new_otu)

    otu_path = await write_otu(new_otu, src_path)

    if otu_path is None:
        logger.error("Failed to write OTU data to directory")
        return

    logger.info("OTU data written to otu_path.", otu_path=str(otu_path))
    logger.warning(
        'otu_path is empty. Use "virtool ref add accessions" before building, updating or submitting.'
    )

    listing_path = await add_new_listing(
        otu_path, catalog_path=catalog_path, logger=logger
    )
    if listing_path is None:
        logger.error("Failed to write listing data to catalog")
        return

    logger.info("Wrote catalog listing", listing=str(listing_path))

    # Sent new OTU path to stdout
    print(str(otu_path))


def generate_otu(taxonomy_data: dict, new_id: str) -> dict:
    """
    Create a new OTU from a given NCBI Taxonomy unique Id (taxid) and
    a pre-generated Virtool OTU Id

    :param taxonomy_data: Dictionary containing fetched NCBI Taxonomy metadata
    :param new_id: Pre-generated Virtool OTU Id
    :return: The contents of a OTU metadata file
    """
    for key in ["ScientificName", "Id"]:
        if key not in taxonomy_data:
            return {}

    otu = {
        "_id": new_id,
        "name": taxonomy_data.get("ScientificName", ""),
        "abbreviation": "",
        "schema": [],
        "taxid": int(taxonomy_data.get("Id", "-1")),
    }

    return otu


async def write_otu(otu: dict, src_path: Path) -> Path:
    """
    Generate a new directory for given OTU metadata, store the metadata under in otu.json
    and return the path to the new directory.

    :param otu: Dict of OTU metadata
    :param src_path: Path to a reference directory
    :return: The new OTU directory path
    """
    dirname = generate_otu_dirname(name=otu["name"], otu_id=otu["_id"])

    otu_path = src_path / dirname
    otu_path.mkdir()

    with open(otu_path / "otu.json", "w") as f:
        json.dump(otu, f, indent=4, sort_keys=True)

    return otu_path
