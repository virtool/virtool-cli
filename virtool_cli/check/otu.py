import json
import re
from pathlib import Path

from structlog import get_logger

from virtool_cli.utils.reference import get_isolate_paths, get_sequence_paths

logger = get_logger("check_otu")


def check_otu(otu_path: Path) -> bool:
    """:param otu_path: A path to an OTU directory under a src reference directory
    :return: True if an OTU's metadata is valid,
        False if a problem is found
    """
    is_valid = True

    try:
        [_, otu_id] = otu_path.name.split("--")
    except ValueError as e:
        logger.exception(e)
        return False

    log = logger.bind(otu_id=otu_id)

    log.debug(f"Running checks on OTU {otu_id}...")

    if not validate_otu(otu_path):
        is_valid = False

    isolate_paths = get_isolate_paths(otu_path)

    if not isolate_paths:
        log.error("No accession data in OTU", otu=otu_path.name)
        is_valid = False
    else:
        for isolate_path in isolate_paths:
            if not validate_isolate(isolate_path):
                is_valid = False

            sequence_paths = get_sequence_paths(isolate_path)

            if sequence_paths:
                if any(
                    not validate_sequence(sequence_path)
                    for sequence_path in sequence_paths
                ):
                    is_valid = False

            else:
                log.error("No sequences in isolate", isolate=isolate_path.name)
                is_valid = False

    if is_valid:
        log.info("OTU is valid")

    return is_valid


def validate_otu(otu_path: Path) -> bool:
    """Checks OTU metadata for bad content and logs a warning if problems are found.

    :param otu_path: the path to an otu directory to validate
    :return: whether the otu is valid
    """
    otu_metadata_path = otu_path / "otu.json"

    try:
        with open(otu_metadata_path) as f:
            otu = json.load(f)
    except FileNotFoundError:
        logger.error("Could not find otu.json")
        return False
    except json.JSONDecodeError:
        logger.error("Could not decode otu.json")
        return False

    is_valid = True

    if "_id" not in otu:
        logger.error("No _id in otu.json")
        is_valid = False

    if "schema" not in otu:
        logger.warning("No schema in otu.json")
        is_valid = False

    if not otu.get("taxid"):
        logger.warning("No taxid in otu.json")
        is_valid = False

    if is_valid:
        logger.info("OTU metadata is valid")

    return is_valid


def validate_isolate(isolate_path: Path) -> bool:
    """Checks isolate metadata for bad content and logs a warning if problems are found.

    :param isolate_path: A path to an isolate directory under an OTU directory
    :return: whether the isolate metadata is valid
    """
    isolate_id = isolate_path.stem
    isolate_json_path = isolate_path / "isolate.json"

    log = logger.bind(id=isolate_id, json_path=str(isolate_json_path))

    try:
        with open(isolate_json_path) as f:
            isolate = json.load(f)
    except FileNotFoundError:
        logger.warning("Could not find isolate.json")
        return False
    except json.JSONDecodeError:
        log.warning("Could not parse isolate.json")
        return False

    is_valid = True

    # Run isolate metadata checks
    if not isolate.get("id", ""):
        log.warning("Invalid or missing id field in isolate.json")
        is_valid = False

    if not isolate.get("source_type", ""):
        log.warning("Invalid or missing source_type field in isolate.json")
        is_valid = False

    if not isolate.get("source_name"):
        log.warning("Invalid or missing source_name field in isolate.json")
        is_valid = False

    if "default" not in isolate or not isinstance(isolate["default"], bool):
        log.error("Invalid or missing default field in isolate.json")
        is_valid = False

    if is_valid:
        log.info("Isolate metadata is valid")

    return is_valid


def validate_sequence(sequence_path: Path) -> bool:
    """Checks a sequence file for bad content and logs a warning if problems are found.

    :param sequence_path: A path to a sequence file under an isolate directory
    :return: whether the sequence metadata is valid
    """
    is_valid = True

    sequence_id = sequence_path.stem

    log = logger.bind(
        sequence_id=sequence_id,
        path=str(sequence_path.relative_to(sequence_path.parents[3])),
    )

    try:
        sequence_data = json.loads(sequence_path.read_text())
    except FileNotFoundError:
        log.error("Missing sequence data")
        return False
    except json.JSONDecodeError:
        log.error("Sequence metadata is not JSON-parseable")
        return False

    accession = sequence_data["accession"]

    # Run metadata checks
    if not sequence_data.get("_id", ""):
        log.error("Sequence metadata lacks a unique Virtool ID in the JSON")
        is_valid = False

    if not verify_accession(accession):
        log.error(
            f"Accession '{accession}' contains invalid characters or capitalization",
        )
        is_valid = False

    if is_valid:
        log.info("Sequence metadata is valid")

    return is_valid


def verify_accession(accession: str) -> bool:
    """Returns True if the accession matches NCBI standards
    The accession must be free of characters other than digits,
    capital letters A-Z, '.', and '_'

    :param accession: An accession to be inspected
    :return: Boolean based on whether the accession is free of invalid characters
    """
    return re.search(r"([^A-Z_.0-9])", accession) is None
