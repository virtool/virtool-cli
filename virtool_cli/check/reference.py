from pathlib import Path

import structlog
from structlog import get_logger

from virtool_cli.check.otu import check_otu
from virtool_cli.utils.reference import get_otu_paths

base_logger = structlog.get_logger()


def check_reference(path: Path) -> bool:
    """Inspects all files in the reference src path and logs a warning for each problem
    found.

    :param path: the path to the reference repository
    :return: whether the reference is valid
    """
    logger = get_logger("check_reference", path=path)
    logger.info("Running checks on reference directory...")

    ref_valid = True

    for otu_path in get_otu_paths(path / "src"):
        if not check_otu(otu_path):
            ref_valid = False
            logger.warning(f"OTU {otu_path.name} has issues", otu=otu_path.name)

    return ref_valid
