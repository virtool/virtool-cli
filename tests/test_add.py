import shutil
import json
import subprocess
from pathlib import Path

import orjson
import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props, paths

from virtool_cli.ref.otu import add_otu

# from virtool_cli.ref.build import build_json
# from virtool_cli.ref.init import init_reference
from virtool_cli.ref.repo import EventSourcedRepo as Repo
from virtool_cli.utils.reference import get_sequence_paths


@pytest.fixture
def empty_repo_path(tmp_path):
    yield tmp_path / "empty_repo"


@pytest.fixture
def precached_repo(empty_repo_path, cache_example_path):
    cache_path = empty_repo_path / ".cache"

    subprocess.run(
        [
            "virtool",
            "ref",
            "init",
            "--data-type",
            "genome",
            "--name",
            "empty repo",
            "--organism",
            "virus",
            "--path",
            str(empty_repo_path),
        ],
        check=False,
    )

    shutil.copytree(cache_example_path, (cache_path / "ncbi"), dirs_exist_ok=True)
    (cache_path / "repo").mkdir(exist_ok=True)

    yield Repo(empty_repo_path)

    shutil.rmtree(empty_repo_path)


def get_all_sequence_paths(otu_path: Path) -> set[Path]:
    return {
        sequence_path
        for isolate_path in otu_path.iterdir()
        if isolate_path.is_dir()
        for sequence_path in get_sequence_paths(isolate_path)
    }


def run_build_reference(repo: Path, output: Path):
    subprocess.run(
        [
            "virtool",
            "ref",
            "build",
            "--path",
            str(repo),
            "--output-path",
            str(output),
        ],
        check=False,
    )


class TestAddOTU:
    @staticmethod
    def run_add_otu(taxid: int, accessions: list[str], path: Path):
        command = [
            "virtool",
            "ref",
            "otu",
            "init",
            "--path",
            str(path),
            "--taxid",
            str(taxid),
        ]

        subprocess.run([*command, *accessions])

    @pytest.mark.parametrize(
        "taxid,accessions",
        [
            (
                438782,
                [
                    "NC_010314",
                    "NC_010315",
                    "NC_010316",
                    "NC_010317",
                    "NC_010318",
                    "NC_010319",
                ],
            ),
        ],
    )
    def test_success(
        self, taxid, accessions, precached_repo, snapshot: SnapshotAssertion
    ):
        self.run_add_otu(taxid, accessions, precached_repo.path)

        run_build_reference(repo=precached_repo.path, output=precached_repo.path)

        built_reference_path = precached_repo.path / "reference.json"

        with open(built_reference_path, "rb") as f:
            built_reference = orjson.loads(f.read())

        for otu in built_reference["otus"]:
            assert otu == snapshot(exclude=props("id", "isolates"))

            for isolate in otu["isolates"]:
                assert isolate == snapshot(exclude=props("id", "sequences"))

                indexed_sequences = {
                    sequence["accession"]: sequence for sequence in isolate["sequences"]
                }

                for accession in sorted(indexed_sequences):
                    assert indexed_sequences[accession] == snapshot(exclude=paths("id"))


@pytest.mark.ncbi
class TestAddOTUAutofill:
    @staticmethod
    def run_add_otu(taxid: int, path: Path):
        subprocess.run(
            ["virtool", "ref", "otu", "create", str(taxid), "--path", str(path)],
            check=False,
        )

    @staticmethod
    def run_build_reference(repo: Path, output: Path):
        subprocess.run(
            [
                "virtool",
                "ref",
                "build",
                "--path",
                str(repo),
                "--output-path",
                str(output),
            ],
            check=False,
        )

    @staticmethod
    def read_build(path) -> dict:
        with open(path, "rb") as f:
            return orjson.loads(f.read())

    @pytest.mark.parametrize("taxid", [438782])
    def test_success(self, taxid, precached_repo: Repo):
        print(list((precached_repo.path / ".cache").iterdir()))
        output_path = precached_repo.path / "reference.json"

        self.run_build_reference(precached_repo.path, precached_repo.path)
        pre_repo = self.read_build(output_path)

        assert pre_repo["otus"] == []

        self.run_add_otu(taxid=taxid, path=precached_repo.path)

        self.run_build_reference(precached_repo.path, precached_repo.path)

        assert output_path.exists()
        post_repo = self.read_build(output_path)

        assert post_repo["otus"]

        for key in ["created_at", "data_type", "name"]:
            assert pre_repo[key] == post_repo[key]

        assert post_repo["otus"] != pre_repo["otus"]
