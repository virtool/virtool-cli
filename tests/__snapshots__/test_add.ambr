# serializer version: 1
# name: TestAddOTU.test_success[438782-accessions0]
  dict({
    'acronym': '',
    'excluded_accessions': list([
    ]),
    'legacy_id': None,
    'molecule': dict({
      'strandedness': 'single',
      'topology': 'circular',
      'type': 'DNA',
    }),
    'name': 'Abaca bunchy top virus',
    'schema': list([
    ]),
    'taxid': 438782,
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].1
  dict({
    'legacy_id': None,
    'name': dict({
      'type': 'isolate',
      'value': 'Q767',
    }),
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].2
  dict({
    'accession': 'NC_010314',
    'definition': 'Abaca bunchy top virus DNA-N, complete genome',
    'legacy_id': None,
    'segment': 'DNA N',
    'sequence': 'AGCAGGGGGGCTTATTATTACCCCCCCTGCTCGGGGCGGGACATTCTGTGATGGGCTGGGCTTTATGCGGCCAAATAAGCCCATAAAGCCAGATCTGGGCCCATTTAAGGGCCCGTGGTTTGAAAATGTCGCGTTCCCGCCTAAATTGTTTGCTTGCCCTGCAAGGAAACGAAAACTCTATAAATAGGGTTGTTCTCTGCTTGTTTAATACATCAGGCGCAAATCTTTTGCAACGATGGATTGGATGGAATCACAATTCAAGACATGTACGCATGGCTGCGACTGGAAGGCGATAGCTCCAGAAGCACAAGATAATATACAGGTAATTACATGTTCCGATTCAGGTTACGGAAGAAAGAACCCTCGTAAGGTTCTTCTGAGGAGTATTCAGATAGGGTTCAATGGAAGCTTCAGAGGAAGTAATAGAAATGTTCGAGGCTTCATATACGTGTCTGTAAGACAGGATGATGGCCAAATGAGACCAATTATGGTCGTTCCATTCGGAGGGTATGGATATCATAACGACTACTATTATTTTGAAGGACAATCCAGTACGAATTGTGAGATAGTGTCGGACTATATTCCGGCCGGTCAAGACTGGAGCAGAGATATGGAGATAAGTATAAGTAACAGCAACAATTGTAATCAAGAGTGCGATATCAAGTGTTATGTAGTATGTAATTTAAGAATTAAGGAATAAWATTGTTGCCGAAGGTCTGTTATTTGAATGTTGAGATAAGGAAAGGGGCGGCGAAGCATGTGTGTATAATAACATATAACACACTATTATATATTTTGTAAAGAATAAAATTATGACCTGTCAGATTAAGTTTAGAATGAACTGAGGCCGAAGGCCTCACCGAGGCCGAAGGCCGTCAGGATGGTTTTACAAAATAATTATAAGCACCTGTACTAAGTACGAAGAGCGGTATAATATCTGAAAGGAAAAAATAATAATATAATAAAAATATTATGATGTCCCAAAATAGCAGAATGCTAAAGGAACAAAAGGATGCTCTAAGTACAGGGTTGCGTGCTCTGGACGCCACTTTAGTGGTGGGCCAGATGTCCCGAGTTAGTGCGCCACGTC',
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].3
  dict({
    'accession': 'NC_010315',
    'definition': 'Abaca bunchy top virus segment 2, complete sequence',
    'legacy_id': None,
    'segment': 'DNA U3',
    'sequence': 'AGCAGGGGGGCTTATTATTACCCCCCCTGCTGGGACGGGACATCCGAATAGTAATGGGCTTAATACATAAATGGGCCGCCAAATAAGCGTACAGTTCAGTATCTTCGTTTTGGGCCTCGGCCCAAAATTAAGAGAACGTGTGCGCTTGTTTGGGTTGGACCGTAGGTCCGGTTCGAATGAGGAATATTCCTTGCTTCCATAGGAAGGAACAGCGAAATAAAATAAATCGCTGTTACACGTGTAAGAATACTATTAGTTCGCAGAGTTTATAAATACCCTTCAAGTTGAAGGGTGTGGTGCTCTCTCTCTTCTGTCAGACGGTGTGCCATGATGAAGCTCTCCGGGAGGAAAGAAGGGCGGAGGTGGAGGAGCCGCCTCGCCCGTCCCTCCGACGAAGTATACGAAGTCGTAGGTATGTCAGGGTATTTATAGAAGCGTCGTATACTAAACGCGCCTGTATCATAATTGTATTTGTGTATTTAAATATTTAAATACCAAACCTTAAGGAATAAAATATATGTTTAATAAAATAACACATGACAATAAAAGAAAACAATAATACAATAATAAAATTGTATTATGCAAACGTTGTATAATTAAGGTCTGTTTAATATACATATATATATGTATATTACAGTATTGTTTAAAATAAATGACTTGGAAGGAAATAATAATATTAATGATAAGCAATAAAATATTCCGAATACATAAAAAGGGGAAAATGCAATTAAGACCTGTCACGTTTACATATTTCGAATATATTAAGGGCCGAAGGCCCGTCAGTATGCAGGTACATCAGTGATTGCTTCGTGACGAAGCAAGGGTATTATTGTAAATAAGAAAGCATCTGACAACTTTAATAGTGGTCCCCTATACAGCTGTCATATGACAGCTGGCAAAGGATCATTGGGCGGACTCCAAATATATATTAAATATAACATATAAAATATATAAGGTATAGATACTATTATTAATTAAAAGAGGCGGGAAAGAGGACACGTGCGCCGCACGCCACTTTAGTGGTGGGCCATATGTCCCGAGTTAGTGCGCTTACGTC',
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].4
  dict({
    'accession': 'NC_010316',
    'definition': 'Abaca bunchy top virus DNA-S, complete genome',
    'legacy_id': None,
    'segment': 'DNA S',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATTCTGTGATGGGCTGGGCTTTATGCGGCCAAATAAGCCCATAAAGCCAGATCTGGGCCCATTTAAGGGCCCGTGGTTTGAAAATGTCGCGTTCCCGCCAAAATAGTTGCTTGCTCTGCAAGCAAACTATATCTATTATAAATACCAGATGTAATGGTTGCCTGAGAATGAAAATAAAAGGATGGCGAGGTATCCGAAGAAGGCGCTAAAGAAGAGGAAGGCGGTACGCCGTAAGTATGGAAGTAAAGCTACGACCAGTCATGATTATGCTGTTGATACTTCATTTATTGTTCCTGAAAATACTATTAAGCTGTATCGTATTGAGCCTACTGATAAAACATTACCCAGATATTTTATCTGGAAAATGTTTATGTTGTTGGTATGCAAAGTCAGACCAGGGCGTATACTTCATTGGGCGATGATTAAGAGCTCTTGGGATGTGAAGGATCCAACAGTTGTTCTTGAAGCGCCTGGATTGTTTATCAAGCCAGCAAACAGCCATCTGGTTAAACTGGTATGCAGTGGCGAGTTAGAAGCTCCAGTAGGAGGAGGGACTTCAGAGGTTGAATGTCTTCTACGGAAGACAACTTTACTTCGTAGGAATGTTACAGAATTGGATTTCTTGTATTTGGCGTTTTATTGTTCTTCTGGAGTTACAATCAACTACCAGAACAGGATTACATATCATGTATAAACCACATAAAATAAATGTGGTGTTGCAGGCATGGGAAGAATAAAACAATGTTTGCCTACGAAATATTTTGGTAAAGTGAAATTATGACCTGTCAGAATTAAGTTTAGAATGAACTGAGGCCGAAGGCCTCACCGAGGCCGAAGGCCGTCAAGTTGGATGAATAAAATACAAGGTATAAGTACGAAGAGCGGTATAATATCTGAAAGGAAATAATAATAATATAATAAAATATTATGATGTCCCAAAATAGCAGAATGCTAAAGGAACAAAAGGATGCTCTAAGTACAGGGTTGCGTGCTCTGGACGCCACTTTAGTGGTGGGCCAGATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].5
  dict({
    'accession': 'NC_010317',
    'definition': 'Abaca bunchy top virus DNA-M, complete genome',
    'legacy_id': None,
    'segment': 'DNA M',
    'sequence': 'GGGGCTGGGGCTTATTATTACCCCCAGCCCCGGAACGGGACATCACGTGCCTCCCTGTACACGCACGTGACATCAAATGGCTTGCTGYGCAAGATAGAAAGGCTTCCTGCGGAAGCCAGGCATCATATATGGGTTGTAGAGGATCAGCCGACTCTATAAATATAGGGAGGCGTGGTCATGGGATTGACAGGAGAACGAGTGAAACAATTCTTTGAATGGTTTCTGTTCTTCTCAGCAATATTTGTTGCAATAACAATAATATATATATTGCTGGCAGTTCTCCTTGAACTACCGAAGTATATTAAGGGTTTAGTACAATACGTAGTGGAGTATATTACTAGACGACGGGTATGGACACGGAGGACCCAATTGACAGAGGCAACCGGAGGAGGCGATATAGAAGCTGTAGGGCATGACAGTCAGGCGTATACGCATACTGTTATGCCATSTGTTCCACCAGTTAGCGCTCCTATATCAAATAGGAGAGCTGATCAGCCTCTTCGATCAAGCGCCGGACCAATGTTTTAAATACCCGTGATATTTAATATGCAAGTGTATAAATACCCATTGTAGATCTGTTTGTAACCTGAATATGCAAAGTATATAATACTTTGTTGTAATGTATAAGTACATTAATAATATACGAAGTATAATGTTGATGCGATGTCTTCGGAAAATGAAGTATACCCAAATACACAAAAAAACATATATGTGGTGTATACTTATTGTTAAGTATAATAAAATTATAATACAAACAAATATATTGTGTATTATAATACATAAAAGAAGACAGAGCTGTGAAGTTAAGTAAGAAAGCGACGGATTCGTATTGGATAATATGATTCGCGGAGCATTACTTAACGGCGAAGTAAAGCATCAGACAAGAATATGACAGCTGTCATATCAACTAAAAAGCATAGCTTGCCGAGCACGCTATGCAATATAAGGGAAAAATGAAATAATGACCTGTCACGTTTACACTTTTCGTAAAGATGTAGGGCCGAAGGCCCTAATGACGCGTGTCATATTCTCTATAGTGGTGGGTCATATGTCCCGAGTTAGTGCGCCACGTAA',
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].6
  dict({
    'accession': 'NC_010318',
    'definition': 'Abaca bunchy top virus DNA-C, complete sequence',
    'legacy_id': None,
    'segment': 'DNA C',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATCACGTGCGCCTATAAAGAAGCACGTGATGATGTGACATATGTTTGCCGTAAGCTGTACAAAAGCGAATGCTTTATTGCTTTATTCGTTGCTTAGCCCCTAAGCAACCTGGTAGTGGGATGTGGTCCCTGATGCGAAGCAGCTTTGTTGTTGCTATTTATATCTGTGTTCGCCATGTAAAATGCGAAATACAACGCTGATCAGAATGGAGTTCTGGAATTCGGAAGCATTTTGCGACGATGTGAAGCGTGTCATTAAGCAAAAATATTGGGAAGAGCGGATGAAATCTCTATTTATAGAGAAGGTGAGTGGTTATGTTCGAAGGATTCTTGTTTATGGTAATCTTGATGATACCATATATGCGGTTCAGCAAATGAAGACTTCTATAGTTCAGTGTGCTGAACGTTTCGGTAAAGCCTGTGTGGTGGTATATAATGGTTTGGATCCATCAATAGGTTTCCGATTACACACTATGGCGTTCTTCTTCGAAGAATATGTTGAGGAAGTGAGTACTGCAGATCCAATGGCGGTTCAGTTATTTTGTGATGAAGAAATAGAAGAATTCTCAAATTCTGATGTACGCCTTATTAAAAATGTTATTATGGCATCGACAGATGCATCGATTGATGTAGGAAATTGTATTCAGATAATAATATCTGATAATGTAATAACATTCTATATATGTTAACTTCATTTATACATAAGAATGAATGAAGTGGTTTATTTATGATTTTTAGAATATAATCATAAATGGTAAACCTTAAGCAAAACCACCCTAAAACAAATAAACCTCTCTGACATACAACCCTCTATAAAATAAAGCCCATGTAAGATTCAAATTTAGAATGAAAAATGGGCCGAAGGCCCATATAAATGCATTTAAGGCCGAAGGCCTTATAATTGCAGGAAGAAAAGAACACGGTTTTGCTTACGTGGCCTGTGGGCCATATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
# name: TestAddOTU.test_success[438782-accessions0].7
  dict({
    'accession': 'NC_010319',
    'definition': 'Abaca bunchy top virus DNA-R, complete genome',
    'legacy_id': None,
    'segment': 'DNA R',
    'sequence': 'GGCAGGGGGGCTTATTATTACCCCCCCTGCCCGGGACGGGACATTTGCATCTATAAATAGAAGCGCCCTCGCTCAACCAGATCAGGCGCTGCAATGGCTAGATATGTCGTATGTTGGATGTTCACCATCAACAATCCCGAAGCTCTTCCAGAGATGAGGGAAGAATACAAATACCTGGTTTACCAGGTGGAGCGAGGCGAAAGCGGTACACGACATGTGCAGGGCTATGTTGAAATGAAGAGACGAAGTTCTCTGAAACAAATGAGGGCTTTAATTCCTGGTGCCCATCTCGAAAAGAGAAGGGGCACACAGGAAGAAGCTAGAGCTTATTGTATGAAGGCAGATACGAGAGTCGAAGGTCCCTTCGAGTTTGGTCTTTTCAAAGTATCATGTAATGATAATTTGTTTGATGTCATACAGGATATGAGAGAAACGCACAAACGGCCGATTGAGTATTTATACGACTGTCCTAATACCTTCGATAGAAGTAAGGATACATTATACAGGGTACAAGCGGAAATGAATAAAATGCAAGCTATGATGTCGTGGTCGGAAACCTATGGTTGCTGGACGAAGGAAGTGGAGGAACTAATGGCGGAGCCATGTCACCGACGGATTATTTGGGTCTATGGCCCAAATGGTGGTGAAGGTAAAACAACCTATGCGAAGCATCTAATCAAGACCAGAAATGCATTTTATACACCTGGCGGAAAGACACTGGATATATGTAGGCTGTATAATTATGAGGGAATTGTAATATTTGATATTCCCAGATGCAAAGAGGATTACTTGAATTACGGAATTCTTGAGGAATTCAAGAATGGCATCATTCAGAGCGGGAAATATGAACCAGTTTTAAAAATTGTAGAGTATGTGGAGGTCATTGTCATGGCTAACTTCCTGCCGAAGGAAGGAATATTCTCGGAAGACCGAATAAAGCTTGTAACTTGTTGAACACGCTATGCAATAAAGGGGAAAAATGCAATTATGACCTGTCACGTTTACACTTTTCGTAAAGATGTAGGGCCGAAGGCCCTAATGACGCGTGTCATATTCTCTATAGTGGTGGGTCATATGTCCCGAGTTAGTGCGCCACGTG',
  })
# ---
